### PR TITLE
Spec Changes Trie

### LINKED ISSUE #129: Spec Storage changes trie
Storage change trie root is part of digest 
   https://github.com/paritytech/substrate/blob/master/test-utils/runtime/src/system.rs#L217
   
   And we need to spec how it is generated.

### PR DESCRIPTION
**Pull request by lamafab**
Requested spec for #129.

@drskalman @FlorianFranzen  Putting the entire behavior and steps of building the Changes Trie into words is kind of hard, so please leave feedback if something isn't clear or needs to be clarified/expanded.

UPDATE (2020-03-10) This was adjusted: ~~Also, in the code there's this special case of `NO_EXTRINSIC_INDEX`, I'll follow up with that (either in this PR or a new one).~~

UPDATE (2020-04-15) Big restructuring based on the feedback I've received. Clarified some sections and added examples. Current Screenshot:

image

### CODE REVIEW

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]: Do you mean the main trie or the changes trie here? Please clarify.

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]:
```suggestion
  encoding for varying data types, this structure starts its indexing at
```

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]:
```suggestion
    storage key and child changes Trie (<reference|defn-storage-key-to-child-tries>)>|<cell|<em|KeyIndex>
```

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]: Please clarify here what exactly is this 32bit number (e. g. it's not completely clear if it's the hash of the header or the state root hash).

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]:
```suggestion
    the Runtime directly, have a <verbatim|0xffffffff> value as their index.
```

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]:
```suggestion
    The key is inserted into the Trie, where its value is a SCALE encoded
```

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]:
```suggestion
    <verbatim|:changes_trie> to storage, where its value is made out of two
```

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]:
```suggestion
    main Trie, where its value is the Merkle proof as desribed in section
```

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]:
```suggestion
    which stores the mappings between the storage key and extrinsics.
```

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]:
```suggestion
  database by the Polkadot node. The Merkle proof of the Changes Trie must be
```

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (lamafab) [reply runtime-environment-spec/polkadot_re_spec.tm]: The Changes Trie. Spec adjusted.

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (lamafab) [reply runtime-environment-spec/polkadot_re_spec.tm]: It's just the block number. Spec adjusted.

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]:
```suggestion
    changes to the key. The indices are unsigned 32-bit integers and its
```

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]: It's not quite clear to me what is meant by generation of the entries. I may try to guess or ask you, but as it's not quite clear now, maybe you might clarify it a bit.

File: runtime-environment-spec/polkadot_re_spec.tm
Comment (Harrm) [inline runtime-environment-spec/polkadot_re_spec.tm]: occur/occured?

Comment (lamafab) [reply runtime-environment-spec/polkadot_re_spec.tm]: -> Any changes that happened before `ext_storages_changes_root` gets called. I'm expanding a section in the storage state transition which explains how the changes get tracked, I'll push it today.

Comment (Harrm) [reply runtime-environment-spec/polkadot_re_spec.tm]: I mean, there's a typo, but still thank you.

**Review by FlorianFranzen — CHANGES_REQUESTED** [review]
Looks good. I rewrote the paragraph before table 2.1 a little bit and added other minor fixes (like updating the TOC). The changes are available in a separate pull request to this branch, see #160.

**Comment by lamafab** [comment]
> Looks good. I rewrote the paragraph before table 2.1 a little bit and added other minor fixes (like updating the TOC). The changes are available in a separate pull request to this branch: w3f/polkadot-spec/pull#160

Yes, I saw that. I'll look into it.

**Comment by drskalman** [comment]
@lamafab could you move it to c02-state.tm?

File: polkadot-host-spec/c02-state.tm
Comment (drskalman) [inline polkadot-host-spec/c02-state.tm]: What is "almost" here? they seems identical to me.

File: polkadot-host-spec/c02-state.tm
Comment (drskalman) [inline polkadot-host-spec/c02-state.tm]: why it is not just simply a tuple type then? and so if it, I would define it as " the tracking change type is a SCALE Encoded tuple
(id, keyindex) in which id is in table blah and key index is blah.

File: polkadot-host-spec/c02-state.tm
Comment (drskalman) [inline polkadot-host-spec/c02-state.tm]: When we talk about "Changes Trie" I think at some point we need to formally define it as a trie. Is it radix-16 trie for example. Are the nodes have exactly the same structure as the state trie etc. So maybe turning this paragraph into a definition saying what type of tree it is and then the key type and value type etc. 
"Track changes in three different types" what does this algorithmically mean? Like we store the key that is change in a trie node? it is not really clearly at least from the first read.

File: polkadot-host-spec/c02-state.tm
Comment (drskalman) [inline polkadot-host-spec/c02-state.tm]: For all these 3 definitions I would define a formal function (give it a name) define its  domain and range type and it is criteria (similar the childrenbitmap for example). It will be less confusing.

File: polkadot-host-spec/c02-state.tm
Comment (drskalman) [inline polkadot-host-spec/c02-state.tm]: It seems that stuff here are so is specifically about ext-storage-changes-root and is about how to basically call that API. Why shouldn't this part be moved into sect-ext-storage-changes-root ?

File: polkadot-host-spec/c02-state.tm
Comment (drskalman) [inline polkadot-host-spec/c02-state.tm]: We don't really talk about blocks before chapter 3. Shouldn't the following paragraph be moved to to where we spec the header of the block and then refer back to this section?

**Review by drskalman — CHANGES_REQUESTED** [review]

File: polkadot-host-spec/c02-state.tm
Comment (lamafab) [reply polkadot-host-spec/c02-state.tm]: Well, that enum ("varying data type"), which is the key, indicates the type of the *value*, not the key itself.

We define "varying data type" as:
image

In this case it's:
image

I guess we can just leave it at "it's a varying data type" and just explicitly say that it refers to the type of the other value, not the key itself.

File: polkadot-host-spec/c02-state.tm
Comment (lamafab) [reply polkadot-host-spec/c02-state.tm]: See [here](https://github.com/w3f/polkadot-spec/pull/142#discussion_r404836931).

File: polkadot-host-spec/c02-state.tm
Comment (lamafab) [reply polkadot-host-spec/c02-state.tm]: I'll reference the radix-16 trie in the Changes Trie section.

> "Track changes in three different types" what does this algorithmically mean?

Depending on the enum, different value types gets inserted into that trie. That type is identified by the enum index. The mappings are:

```
<changed-key> --> <array-of-extrinsic-indexes>
<changed-key> --> <array-of-block-numbers>
<changed-key> --> <array-of-child-changes-tries>
```

I'll change the "tracks changes *in*..." to "tracks changes *of*" this as:

> The Changes Trie tracks changes of three different types of mappings (or search indices) which are differentiated by their key as defined in table 2.1.

I'm also replacing "Followed by" with "Value":
image

File: polkadot-host-spec/c02-state.tm
Comment (lamafab) [reply polkadot-host-spec/c02-state.tm]: This was moved to the `ext_storage_changes_root` section.

File: polkadot-host-spec/c02-state.tm
Comment (lamafab) [reply polkadot-host-spec/c02-state.tm]: As discussed, the Changes Trie section was moved to chapter 3.

File: polkadot-host-spec/c03-transition.tm
Comment (Harrm) [inline polkadot-host-spec/c03-transition.tm]: The meaning of this is not quite clear. For a block where block mappings are generated the changes trie must include entries related to any extrinsic changed in the tracked range of blocks, right? What does per-block basis mean in this case?

File: polkadot-host-spec/ae-hostapi.tm
Comment (Harrm) [inline polkadot-host-spec/ae-hostapi.tm]: I don't really understand what hierarchy is meant here. Does that mean that if, for example, I have interval of 3 and levels of 4, I have to add to the changes trie of every 3rd block the entries about  changes made in past 3 blocks, to every 9th block changes made in past 9 blocks, to every 27th block for past 27 blocks and so on to 3^4? So I have to store the information about what keys were changed by which blocks for 81 blocks?

File: polkadot-host-spec/c02-state.tm
Comment (Harrm) [inline polkadot-host-spec/c02-state.tm]: Could you clarify  the meaning of "identify any changes made by each block"? Does that mean that there should be a mechanism to get the list of keys changed by a block for any block in the block tree?

File: polkadot-host-spec/ae-hostapi.tm
Comment (Harrm) [inline polkadot-host-spec/ae-hostapi.tm]: Sorry, not quite clear what is a change root.

File: polkadot-host-spec/ae-hostapi.tm
Comment (FlorianFranzen) [reply polkadot-host-spec/ae-hostapi.tm:223]: Well, the nice thing about the way this spacing works is

Comment (FlorianFranzen) [reply polkadot-host-spec/ae-hostapi.tm]: You only need to keep a handful of digest tries around. E.g. for `interval = 3` and `level = 4` the you would have a try over 3, 9, 27, 81 blocks. Whenever one of these intervals has passed, you add the current trie for this interval to the merkel prove. Then the trie of that interval is reset (I assume they are also added to a local database somewhere). If you want to save on computation you could update the higher intervals only when the interval just below it is reset. Because the intervals contain multiples of each other, you can use the already consolidated tries form the lower levels instead of adding each single transaction to each interval trie every single time.

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [inline polkadot-host-spec/c03-transition.tm]: - is "Child-Changes-Trie" is a type? if we are storing the root here then maybe it is better to call it the root of ...
- You might want to remind the reader what is H_r.

Comment (drskalman) [inline polkadot-host-spec/c03-transition.tm]: - is "Child-Changes-Trie" is a type? if we are storing the root here then maybe it is better to call it the root of ...

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [inline polkadot-host-spec/c03-transition.tm]: - I wouldn't call them pairs. I'd call them "nodes".
- It is not clear what is happening here. I'm not sure why we need to have word "consider". It is better state that Host make a different change trie for each block whose nodes only consists of the keys that are changed in that block. (even if substrate is keeping one change trie and do some magic to keep other nodes out of calculation).

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [inline polkadot-host-spec/c03-transition.tm]: This confusing. I'm not sure if it should be here. you haven't given an algorithm to generate the trie yet. So as a reader, I assume that something will go in the trie not everything in the universe goes into the trie. So I don't see the reason for this warning. I'd just say "We describe the algorithm to generate the change trie on each block in section blah".

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [inline polkadot-host-spec/c03-transition.tm]: Reading these my understanding is that type 1 and 3 nodes are computed and inserted into the change trie of each block, while type 2 nodes are inserted in a trie that is generate with specific request of runtime. If that is the case then I wouldn't mention type 2 here. I move it to API section where it generates change tries which also have type 2 nodes.

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [reply polkadot-host-spec/c03-transition.tm]: - again if we are generating try always without type 2 then I wouldn't mention type 2 in this section.

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [inline polkadot-host-spec/c03-transition.tm]: I won't mention the "pair" anymore, I think people knows hash table or a trie stores a (key, value) pair.I don't think they need to be reminded at every step. I'd just refer to its value such as "list of extrisics"  and drop the word pair from here after.

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [inline polkadot-host-spec/c03-transition.tm]: use \STATE instead of putting the line numbers manually

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [inline polkadot-host-spec/c03-transition.tm]: Use \FOR instead

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [reply polkadot-host-spec/c03-transition.tm]: use \IF instead

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [inline polkadot-host-spec/c03-transition.tm]: in my opinion there are too many example here. You might one to drop at least one. But your call.
I would also demonstrate the whole calculation for at least one of the example the exact value for the key (type, ?, Key) and the C_Value/A_value.

File: polkadot-host-spec/c03-transition.tm
Comment (drskalman) [reply polkadot-host-spec/c03-transition.tm]: also in type 2 as we discussed, the B in H_i(B) is the  last block of the interval I think this need to be clarified.

**Review by drskalman — CHANGES_REQUESTED** [review]
OK this is my review for 03-transition.tm I haven't looked at the change trie host api yet. I also made some changes directly. I'm going to commit those as well.

**Comment by drskalman** [comment]
#129 says it is on hold so I'm dropping the waiting-for-review label.

**Comment by lamafab** [comment]
Merging this for now (https://github.com/w3f/polkadot-spec/issues/129#issuecomment-680972945).

Also, added an disclaimer:

image

**Pull request merged**
